core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala

```diff
@@ -312,7 +313,51 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
     timeout.max(member.rebalanceTimeoutMs)
   }
 
-  def generateMemberIdSuffix = UUID.randomUUID().toString
+  def generateMemberId(clientId: String,
+                       groupInstanceId: Option[String]): String = {
+    groupInstanceId match {
+      case None =>
+        clientId + MEMBER_ID_DELIMITER + UUID.randomUUID().toString
+      case Some(instanceId) =>
+        instanceId + MEMBER_ID_DELIMITER + currentStateTimestamp.get
+    }
+  }
+
+  // This validation doesn't handle the case where a static member changes
+  // its group.instance.id while maintains the same member.id. That edge case should
+  // be caused by a hacked consumer client since current client logic doesn't
+  // support group.instance.id change on runtime.
+  def validateMemberIdIfStatic(memberId: String): Boolean = {
+    val delimiterIdx = memberId.lastIndexOf(MEMBER_ID_DELIMITER)
+    // Member id must have formatted with valid delimiter as [prefix]-[suffix] to
+    // be validated.
+    if (delimiterIdx == -1) {
+      info(s"given member.id $memberId was ill-formatted without delimiter '-'.")
```

Comment:
Should this happen under normal case? If not we should log it as ERROR and return false.

Reply:
We are being extra cautious here since we don't want to unexpectedly fence any member without this structure. Returning true means we don't check this case in static member.id validation, but it is not guaranteed to be valid.

Reply:
If we think that client code should not validly set their member.ids, i.e. they should always be set by brokers and broker code always set it in a way of `*-*`, then this should not happen and I think it is okay to treat as a fatal error and reject the client request.